File: python/examples/payload_with_typedata.py
from pathlib import Path

from drift_protocol.common import DriftPackage, DataPayload, StatusCode
from drift_protocol.meta import MetaInfo, TypedDataInfo
from drift_bytes import Variant, InputBuffer, OutputBuffer

from google.protobuf.timestamp_pb2 import Timestamp
from google.protobuf.any_pb2 import Any


def create_typed_package() -> bytes:
    global original
    pb_time = Timestamp()
    pb_time.GetCurrentTime()
    DATA = [Variant(False), Variant(1), Variant(3.14), Variant("Hello World")]
    NAMES = ["bool", "int", "float", "string"]

    # Create a package and serialize it
    original = DriftPackage()
    original.id = pb_time.ToMilliseconds()
    original.source_timestamp.CopyFrom(pb_time)
    original.publish_timestamp.CopyFrom(pb_time)
    original.status = StatusCode.GOOD
    # Fill meta data
    info = TypedDataInfo()
    for name in NAMES:
        var = TypedDataInfo.Item()
        var.name = name
        var.status = StatusCode.GOOD
        info.items.append(var)
    original.meta.type = MetaInfo.TYPED_DATA
    original.meta.typed_data_info.CopyFrom(info)
    # Put data in buffer
    buffer = OutputBuffer()
    for var in DATA:
        buffer.push(var)

    # Prepare payload
    payload = DataPayload()
    payload.data = buffer.bytes()
    msg = Any()
    msg.Pack(payload)
    original.data.append(msg)

    label = DriftPackage.Label()
    label.key = "host_name"
    label.value = "my_host"
    original.labels.append(label)

    label = DriftPackage.Label()
    label.key = "topic"
    label.value = "opcua"
    original.labels.append(label)

    # Serialize package to message
    return original.SerializeToString()


def parse_typed_data(message: bytes) -> None:
    # Parse the package
    new_package = DriftPackage()
    new_package.ParseFromString(message)
    print(
        f"Package ID={new_package.id} type={MetaInfo.DataType.Name(new_package.meta.type)}"
    )

    print("Labels:")
    for label in new_package.labels:
        print(f"{label.key}={label.value}")

    payload = DataPayload()
    new_package.data[0].Unpack(payload)
    buffer = InputBuffer(payload.data)

    print("Values:")
    for i, var in enumerate(new_package.meta.typed_data_info.items):
        print(f"{var.name}={buffer.pop().value}, status={var.status}")


if __name__ == "__main__":
    message = create_typed_package()
    parse_typed_data(message)

File: cpp/examples/payload_with_typedata.cc
#include <drift_protocol/common/data_payload.pb.h>
#include <drift_protocol/common/drift_package.pb.h>
#include <google/protobuf/util/time_util.h>
#include <drift_bytes/bytes.h>

using drift::proto::common::DataPayload;
using drift::proto::common::DriftPackage;
using drift::proto::common::StatusCode;
using drift::proto::meta::MetaInfo;
using drift::proto::meta::TypedDataInfo;

using drift_bytes::Variant;
using drift_bytes::InputBuffer;
using drift_bytes::OutputBuffer;

using google::protobuf::util::TimeUtil;
using google::protobuf::Any;

int main() {
    const auto pb_time = TimeUtil::GetCurrentTime();
    std::string message;

    const auto kData = {Variant(true), Variant(1), Variant(1.0), Variant("Hello")};
    const std::vector<std::string> kNames = {"bool", "int", "float", "string"};

    {
        // Create a package and serialize it
        DriftPackage original;
        original.set_id(TimeUtil::TimestampToMilliseconds(pb_time)); // UNIX timestamp in ms
        original.set_status(StatusCode::GOOD);
        original.mutable_source_timestamp()->CopyFrom(pb_time);
        original.mutable_publish_timestamp()->CopyFrom(pb_time);

        // Fill meta data
        TypedDataInfo info;
        for (const auto &name: kNames) {
            auto var = info.add_items();
            var->set_name(name);
            var->set_status(StatusCode::GOOD);
        }

        MetaInfo meta;
        meta.set_type(MetaInfo::TYPED_DATA);
        meta.mutable_typed_data_info()->CopyFrom(info);

        original.mutable_meta()->CopyFrom(meta);

        // Put data in buffer without decomposition and compression
        OutputBuffer buffer;
        for (const auto &data: kData) {
            buffer.push_back(data);
        }

        // Prepare payload
        DataPayload payload;
        payload.set_data(buffer.str());
        original.add_data()->PackFrom(payload);

        // Serialize package to message
        original.SerializePartialToString(&message);
    }

    // Parse the package
    DriftPackage new_package;
    new_package.ParseFromString(message);
    std::cout << "Package with ID=" << new_package.id() << " type="
              << MetaInfo::DataType_Name(new_package.meta().type()) << std::endl;

    DataPayload payload;
    new_package.data(0).UnpackTo(&payload);
    auto data = payload.data();
    auto buffer = InputBuffer(std::move(data));

    // Print data
    for (const auto &var: new_package.meta().typed_data_info().items()) {
        std::cout << var.name() << ": " << buffer.pop() << std::endl;
    }
}
